File: abc286/E/main.py
#!/usr/bin/env pypy3
from collections import defaultdict

N = int(input())
A = list(map(int, input().split()))
S = []
for i in range(N):
  S.append(input())

dists = []
for i in range(N):
  dists.append([1e18] * N)
vals = []
for i in range(N):
  vals.append([0] * N)

for i in range(N):
  for j in range(N):
    if S[i][j] == 'Y':
      dists[i][j] = 1
      vals[i][j] = A[j]
    else:
      vals[i][j] = A[j]

for i in range(N):
  for j in range(N):
    for k in range(N):
      if dists[j][k] > dists[j][i] + dists[i][k]:
        dists[j][k] = min(dists[j][k], dists[j][i] + dists[i][k])
        vals[j][k] = vals[j][i] + vals[i][k]
      elif dists[j][k] == dists[j][i] + dists[i][k] and vals[j][k] < vals[j][i] + vals[i][k]:
        vals[j][k] = vals[j][i] + vals[i][k]

Q = int(input())
for i in range(Q):
  u,v = map(int, input().split())
  u-= 1
  v -= 1

  d = dists[u][v]
  if d == 1e18:
    print('Impossible')
    continue
  print(d, vals[u][v] + A[u])

File: abc286/E/main.cpp
#include <iostream>
#include <map>
#include <vector>
using namespace std;

const int INF = 1e9;

int N;
vector<int> A;
vector<string> S;
map<pair<int, int>, int> dists;
map<pair<int, int>, int> vals;

int main() {
  cin >> N;
  A.resize(N);
  for (int i = 0; i < N; i++)
    cin >> A[i];
  S.resize(N);
  for (int i = 0; i < N; i++)
    cin >> S[i];

  for (int i = 0; i < N; i++) {
    for (int j = 0; j < N; j++) {
      if (S[i][j] == 'Y') {
        dists[{i, j}] = 1;
        vals[{i, j}] = A[j];
      } else {
        dists[{i, j}] = INF;
        vals[{i, j}] = A[j];
      }
    }
  }

  for (int i = 0; i < N; i++) {
    for (int j = 0; j < N; j++) {
      for (int k = 0; k < N; k++) {
        if (dists[{j, k}] > dists[{j, i}] + dists[{i, k}]) {
          dists[{j, k}] = min(dists[{j, k}], dists[{j, i}] + dists[{i, k}]);
          vals[{j, k}] = vals[{j, i}] + vals[{i, k}];
        } else if (dists[{j, k}] == dists[{j, i}] + dists[{i, k}] && vals[{j, k}] < vals[{j, i}] + vals[{i, k}]) {
          vals[{j, k}] = vals[{j, i}] + vals[{i, k}];
        }
      }
    }
  }

  int Q;
  cin >> Q;
  for (int i = 0; i < Q; i++) {
    int u, v;
    cin >> u >> v;
    u--;
    v--;

    int d = dists.count({u, v}) ? dists[{u, v}] : INF;
    if (d == INF) {
      cout << "Impossible" << endl;
      continue;
    }
    cout << d << " " << vals[{u, v}] + A[u] << endl;
  }
  return 0;
}